x-pack/plugins/ml/server/plugin.ts

```diff
@@ -0,0 +1,170 @@
+import { calendars } from './routes/calendars';
+import { dataFeedRoutes } from './routes/datafeeds';
+import { dataFrameAnalyticsRoutes } from './routes/data_frame_analytics';
+import { dataRecognizer } from './routes/modules';
+import { dataVisualizerRoutes } from './routes/data_visualizer';
+import { fieldsService } from './routes/fields_service';
+import { fileDataVisualizerRoutes } from './routes/file_data_visualizer';
+import { filtersRoutes } from './routes/filters';
+import { indicesRoutes } from './routes/indices';
+import { jobAuditMessagesRoutes } from './routes/job_audit_messages';
+import { jobRoutes } from './routes/anomaly_detectors';
+import { jobServiceRoutes } from './routes/job_service';
+import { jobValidationRoutes } from './routes/job_validation';
+import { notificationRoutes } from './routes/notification_settings';
+import { resultsServiceRoutes } from './routes/results_service';
+import { systemRoutes } from './routes/system';
+
+declare module 'kibana/server' {
+  interface RequestHandlerContext {
+    ml?: {
+      mlClient: IScopedClusterClient;
+    };
+  }
+}
+
+export class MlServerPlugin {
+  private readonly pluginId: string = PLUGIN_ID;
+  private log: Logger;
+  private version: string;
+
+  private licenseCheckResults: LicenseCheckResult = {
+    isAvailable: false,
+    isActive: false,
+    isEnabled: false,
+    isSecurityDisabled: false,
+  };
+
+  constructor(ctx: PluginInitializerContext) {
+    this.log = ctx.logger.get();
+    this.version = ctx.env.packageInfo.branch;
+  }
+
+  public setup(coreSetup: CoreSetup, plugins: PluginsSetup) {
+    let sampleLinksInitialized = false;
+
+    plugins.features.registerFeature({
+      id: 'ml',
+      name: i18n.translate('xpack.ml.featureRegistry.mlFeatureName', {
+        defaultMessage: 'Machine Learning',
+      }),
+      icon: 'machineLearningApp',
+      navLinkId: 'ml',
+      app: ['ml', 'kibana'],
+      catalogue: ['ml'],
+      privileges: {},
+      reserved: {
+        privilege: {
+          savedObject: {
+            all: [],
+            read: [],
+          },
+          ui: [],
+        },
+        description: i18n.translate('xpack.ml.feature.reserved.description', {
+          defaultMessage:
+            'To grant users access, you should also assign either the machine_learning_user or machine_learning_admin role.',
+        }),
+      },
+    });
+
+    // Can access via router's handler function 'context' parameter - context.ml.mlClient
+    const mlClient = coreSetup.elasticsearch.createClient('ml', {
+      plugins: [elasticsearchJsPlugin],
+    });
+
+    coreSetup.http.registerRouteHandlerContext('ml', (context, request) => {
+      return {
+        mlClient: mlClient.asScoped(request),
+      };
+    });
+
+    const routeInit = {
+      router: coreSetup.http.createRouter(),
+      getLicenseCheckResults: () => this.licenseCheckResults,
+    };
+
+    annotationRoutes({
+      ...routeInit,
```

Comment:
i wonder if it's worth merging these into a single object?
An alternative approach could be to pass the `routeInit` as a single argument to each route, so they are the same, and for the routes that require additional information, pass them as a second argument.
The routes themselves could then manage the type of that second "additional dependancies" parameter.
Something like:
```
systemRoutes(routeInit, {
      spacesPlugin: plugins.spaces,
      cloud: plugins.cloud,
    });
```

Reply:
Good idea - used the second idea of having an 'additional dependencies' second parameter where applicable. 609be5c4c218d1feed4b249158c438ce7c30f99d